lib/charms/layer/caas_base.py

```diff
@@ -3,26 +3,43 @@
 from pathlib import Path
 from subprocess import run, PIPE, CalledProcessError
 import re
+import tempfile
 
 import yaml
 from charmhelpers.core.hookenv import log
 
 
-def pod_spec_set(spec):
+def pod_spec_set(spec, k8s_resources=None):
+
+    def run_cmd(cmd_args, std_in):
+        log(f'pod-spec-set {cmd_args}, \n{std_in}', level='ERROR')
+        try:
+            run(cmd_args, stdout=PIPE, stderr=PIPE, check=True, input=std_in)
+        except CalledProcessError as err:
+            stderr = err.stderr.decode('utf-8').strip()
+            log(f'pod-spec-set encountered an error: `{stderr}`', level='ERROR')
+
+            if re.match(r'^ERROR application [\w-]+ not alive$', stderr):
+                log('Ignored error due to pod-spec-set getting called during app removal', level='INFO')
+                return
+            raise
+
     if not isinstance(spec, str):
         spec = yaml.dump(spec)
 
-    try:
-        run(['pod-spec-set'], stdout=PIPE, stderr=PIPE, check=True, input=spec.encode('utf-8'))
-    except CalledProcessError as err:
-        stderr = err.stderr.decode('utf-8').strip()
-        log(f'pod-spec-set encountered an error: `{stderr}`', level='ERROR')
+    cmd_args = ['pod-spec-set']
+    if k8s_resources is None:
+        return run_cmd(cmd_args, spec.encode('utf-8'))
 
-        if re.match(r'^ERROR application [\w-]+ not alive$', stderr):
-            log('Ignored error due to pod-spec-set getting called during app removal', level='INFO')
-            return
+    # include k8s_resources.
+    if not isinstance(k8s_resources, str):
+        k8s_resources = yaml.dump(k8s_resources)
 
-        raise
+    fp = tempfile.NamedTemporaryFile(delete=False)
```

Comment:
pod-spec-set can either take a --file arg and now also a --k8s-resources arg.
We also allow stdin to be used. If we are using stdin, we should accept a concatenated spec comprising bot h the generic and k8s stuff. We don't need  the tmp file - just add together the 2 strings and pass through to stdin when calling run_cmd